### ISSUE TITLE
Instructions for "Writing Your First Application" outdated

### ISSUE THREAD

**Issue by Ionstream** [PX4/PX4-user_guide#1335]
Followed the instruction here for `px4_simple_app`, ~~but it will not allow you to build a working program. Code as per docs:~~ but was surprised when it decided to build today (EDIT: I was dumb enough to try using `//` instead of `#` in a `CMakeLists.txt` file). Docs require some updating to match code in `PX4-Autopilot:master`
### Code as per docs
```
/**
 * @file px4_simple_app.c
 * Minimal application example for PX4 autopilot
 *
 * @author Example User <mail@example.com>
 */

#include <px4_platform_common/log.h>

__EXPORT int px4_simple_app_main(int argc, char *argv[]);

int px4_simple_app_main(int argc, char *argv[])
{
	PX4_INFO("Hello Sky!");
	return OK;
}
```

### Code as per src/examples/px4_simple_app
```
/**
 * @file px4_simple_app.c
 * Minimal application example for PX4 autopilot
 *
 * @author Example User <mail@example.com>
 */

#include <px4_platform_common/px4_config.h>
#include <px4_platform_common/tasks.h>
#include <px4_platform_common/posix.h>
#include <unistd.h>
#include <stdio.h>
#include <poll.h>
#include <string.h>
#include <math.h>

#include <uORB/uORB.h>
#include <uORB/topics/vehicle_acceleration.h>
#include <uORB/topics/vehicle_attitude.h>

__EXPORT int px4_simple_app_main(int argc, char *argv[]);

int px4_simple_app_main(int argc, char *argv[])
{
	PX4_INFO("Hello Sky!");

	/* subscribe to vehicle_acceleration topic */
	int sensor_sub_fd = orb_subscribe(ORB_ID(vehicle_acceleration));
	/* limit the update rate to 5 Hz */
	orb_set_interval(sensor_sub_fd, 200);

	/* advertise attitude topic */
	struct vehicle_attitude_s att;
	memset(&att, 0, sizeof(att));
	orb_advert_t att_pub = orb_advertise(ORB_ID(vehicle_attitude), &att);

	/* one could wait for multiple topics with this technique, just using one here */
	px4_pollfd_struct_t fds[] = {
		{ .fd = sensor_sub_fd,   .events = POLLIN },
		/* there could be more file descriptors here, in the form like:
		 * { .fd = other_sub_fd,   .events = POLLIN },
		 */
	};

	int error_counter = 0;

	for (int i = 0; i < 5; i++) {
		/* wait for sensor update of 1 file descriptor for 1000 ms (1 second) */
		int poll_ret = px4_poll(fds, 1, 1000);

		/* handle the poll result */
		if (poll_ret == 0) {
			/* this means none of our providers is giving us data */
			PX4_ERR("Got no data within a second");

		} else if (poll_ret < 0) {
			/* this is seriously bad - should be an emergency */
			if (error_counter < 10 || error_counter % 50 == 0) {
				/* use a counter to prevent flooding (and slowing us down) */
				PX4_ERR("ERROR return value from poll(): %d", poll_ret);
			}

			error_counter++;

		} else {

			if (fds[0].revents & POLLIN) {
				/* obtained data for the first file descriptor */
				struct vehicle_acceleration_s accel;
				/* copy sensors raw data into local buffer */
				orb_copy(ORB_ID(vehicle_acceleration), sensor_sub_fd, &accel);
				PX4_INFO("Accelerometer:\t%8.4f\t%8.4f\t%8.4f",
					 (double)accel.xyz[0],
					 (double)accel.xyz[1],
					 (double)accel.xyz[2]);

				/* set att and publish this information for other apps
				 the following does not have any meaning, it's just an example
				*/
				att.q[0] = accel.xyz[0];
				att.q[1] = accel.xyz[1];
				att.q[2] = accel.xyz[2];

				orb_publish(ORB_ID(vehicle_attitude), att_pub, &att);
			}

			/* there could be more file descriptors here, in the form like:
			 * if (fds[1..n].revents & POLLIN) {}
			 */
		}
	}

	PX4_INFO("exiting");

	return 0;
}
```

**Comment by hamishwillee** [comment]
Thank you. Would you like to try do a PR for an update?

**Comment by Ionstream** [comment]
Give me some time to sort this out, I'm still trying to wrap my head around the starting code.

**Comment by Ionstream** [comment]
If it's okay with you, I propose stretching out the "Gettting Started" section. The learning curve is rather steep, and there's quite a bit of code for new or beginning developers to take in, especially in the subsections "Writing Your First Application" and "Application/Module Template". I also feel that it's easy to get lost on certain pages, since there are too many categories and bits of information to parse through at once. Here's what I have in mind for the first part:

### 1. Writing Your First Application
#### a) "Hello Sky!" - Code
This section introduces PX4 libraries/dependencies, the `__EXPORT` keyword `PX4_INFO()` console logger,  through a simple application that prints out "Hello Sky!" and "Goodbye!" when run. It also helps to familiarise users with the structure of `PX4-Autopilot`.

#### b) "Hello Sky!" - Build
This section introduces the *build system* in PX4 through the elements `CMakeLists.txt` and `px4_add_module()` that are required compile source code into an executable to be loaded together as new firmware.  It also helps to familiarise users with the structure of `PX4-Autopilot`.

#### c) "Hello Sky" - Upload
This section introduces the tools required to flash/upload the new firmware onto PX4-supported hardware by introducing the `upload` parameter and bootloader flashing instructions (if required). It also familiarises users with the buzzer system for identifying hardware states and troubleshooting.

#### d) "Hello Sky" - Run
This section introduces the `MAVLink Shell` that can be used to interface with PX4 hardware through a command-line interface. Users will learn how to run their app/module in `nsh>` and access basic help commands.

**Comment by hamishwillee** [comment]
@Ionstream Thanks very much. I mostly work on PX4 stuff on Weds/Thurs, so can't "properly" support you on this until next week. 

That said I suggest we do this in two parts. The first one being to just update this document so that it matches the current structure (but "works") and get that PR in. Then we look at restructure as a post process. 

With respect to the broader question of the getting started section it depends on what changes you want to make. The developer setup and building PX4 are "fairly" close to right in my opinion - though we should pull out everything after [troubleshooting here](http://docs.px4.io/master/en/dev_setup/building_px4.html) into its own doc. I will do that at some point.

With respect to specifically the writing application and module template, what you are suggesting sounds reasonable - and after all you are actually experiencing the problems now, so a good person to comment and attempt to fix them. In particular the module template reads as though it builds on the "writing .." doc, but feels insufficient. 

But as I say, let's just get this "not broken" first. Make sense?

**Comment by Ionstream** [comment]
Sure! Acutally, there wasn't anything wrong with the code; I managed to nail it down to the "//" I used for comments, which works for C++ but not for C (noob here, sorry :P). In light of this there isn't any update required per se; apologies about that. I think we can close this issue.

That said, I'll start a PR adding some explanations/comments for the minimal application. I think it will be helpful as an entry point to inexperienced programmers like myself.
